docs/proposals/19-migrating-to-prometheus-operator.md.md

```diff
@@ -0,0 +1,444 @@
+       to discover these objects.
+
+    ```yaml
+    serviceMonitorSelector:
+      matchLabels:
+        cluster: shoot--project--name
+    ```
+
+    1. Prometheus needs to discover targets running in the shoot cluster.
+       Normally, this is done by changing the `api_server` field in the config
+       ([example][apiserver-example]). This is currently not possible with the
+       prometheus operator; however, there is an open [issue][prom-op-issue].
+
+        - Alternative 1: Use [additional scrape config]. In this case, the
+          prometheus config snippet is put into a secret and the
+          [prometheus-operator] will append it to the config. The downside here is
+          that it is only possible to have 1 `additional-scrape-config` per
+          Prometheus. This could be an issue if multiple components will need to
+          use this.
+
+        - Alternative 2: A second prometheus can be created that is running in
+          [agent mode]. This prometheus can also be deployed/managed by the
+          [prometheus-operator]. The agent prometheus can be configured to use the
+          API Server for the shoot cluster and use service discovery in the
+          shoot. The metrics can then be written via remote write to the
+          "normal" prometheus or federated.
+
+        - Alernative 3: Nothing in the shoot cluster will be monitored. This is
+          not really an alternative, but perhaps a stopgap solution until either
+          of the other solutions have been implemented.
+
+    1. Deploy an optional [alertmanager][shoot-alertmanager] that is deployed
+       whenever the end-user [specifies][alerting-doc] alerting.
+
+        - Create an `Alertmanager` resource
+
+        - Create the `AlertmanagerConfig`
+
+    1. Health checks - The gardenlet periodically checks the status of the
+       Prometheus `StatefulSet` among other components in the shoot care
+       controller. The gardenlet knows which component to check based on labels.
+       Since the gardenlet is no longer deploying the `StatefulSet` directly and
+       rather a `Prometheus` resource, it does not know which labels are
+       attached to the Prometheus `StatefulSet`. However, the
+       [prometheus-operator] will create `StatefulSets` with the same labelset
+       that the `Prometheus` resource has. The gardenlet will be able to
+       discover the `StatefulSet` because it knows the labelset of the
+       `Prometheus` resource.
+
+    1. Many of the components that require monitoring in the shoot control plane
+       are also covered by the [kube-prometheus]. For example,
+       `kube-state-metrics` and `kube-controller-manager` configuration is
+       already provided by the community. Gardener could adjust and apply the
+       manifests provided by this project. How exactly this should be done is
+       not clear yet, but if it is possible Gardener will have to manage less
+       configuration and can rely on community efforts.
```

Comment:
That would be awesome if we could reuse these configs!
I think one option would be to copy the manifests during code generation and then manipulate and deploy them via code.